clients/pkg/promtail/targets/docker/target.go

```diff
@@ -185,34 +193,50 @@ func (t *Target) process(frames chan []byte, logStream string) {
 			ts = curTs
 		}
 
+		// If time has changed, we are looking at a new event (although we should have seen a new line..),
+		// so flush the buffer if we have one.
+		if ts != curTs {
+			discardRemainingLine = false
+			if payloadAcc.Len() > 0 {
+				t.handleOutput(logStream, curTs, payloadAcc.String())
+				payloadAcc = new(strings.Builder)
+			}
+		}
+
+		// Check if we have the end of the event
+		var isEol = strings.HasSuffix(payload, "\n")
+
+		// If we are currently discarding a line (due to size limits), skip ahead, but don't skip the next
+		// frame if we saw the end of the line.
+		if discardRemainingLine {
+			discardRemainingLine = !isEol
+			continue
+		}
+
+		// Strip newline ending if we have it
+		payload = strings.TrimRight(payload, "\r\n")
+
 		// Fast path: Most log lines are a single frame. If we have a full line in frame and buffer is empty,
 		// then don't use the buffer at all.
-		if payloadAcc.Len() == 0 && strings.HasSuffix(payload, "\n") {
+		if payloadAcc.Len() == 0 && isEol {
 			t.handleOutput(logStream, ts, payload)
 			continue
 		}
 
-		// If time has changed, we are looking at a different event (although we should have seen a new line..),
-		// so flush the buffer.
-		if payloadAcc.Len() > 0 && ts != curTs {
-			t.handleOutput(logStream, curTs, payloadAcc.String())
-			payloadAcc = new(strings.Builder)
-		}
 		// Add to buffer
 		payloadAcc.WriteString(payload)
 		curTs = ts
+
 		// Send immediately if line ended or we built a very large event
-		if strings.HasSuffix(payload, "\n") || payloadAcc.Len() >= softMaxBuffer {
+		if isEol || payloadAcc.Len() > sizeLimit {
+			discardRemainingLine = !isEol
 			t.handleOutput(logStream, curTs, payloadAcc.String())
 			payloadAcc = new(strings.Builder)
 		}
 	}
 }
 
 func (t *Target) handleOutput(logStream string, ts time.Time, payload string) {
-	// We don't output trailing newlines
-	payload = strings.TrimSuffix(payload, "\n")
-	payload = strings.TrimSuffix(payload, "\r")
```

Comment:
should we be removing these?

Reply:
I moved this into the `process` function which now has a `TrimRight` call for these. The reason is that `process` keeps track of the length of the buffer, to see if it's over the new limit we set, but this length should exclude the trailing newline (because we don't output the newline). And I thought it was just easier to take away the trailing newline earlier in the process for this purpose (otherwise we'd have to keep track of the buffer length in a more complicated manner than just calling `payloadAcc.Len()`).